File: templates/python/interfaces/iagent_discovery.py
"""
Discovery layer interface.
Swap the implementation to change the backend:
  - LocalDiscovery    → in-memory (dev / testing)
  - HttpDiscovery     → REST-based registry
  - GossipDiscovery   → P2P gossip protocol
  - OnChainDiscovery  → ERC-8004 on-chain registry
"""

from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import List, Optional


@dataclass
class NetworkInfo:
    protocol: str   # "http" | "websocket" | "grpc" | "tcp"
    host: str
    port: int
    tls: bool = False
    peer_id: str = ""
    multiaddr: str = ""


@dataclass
class HealthStatus:
    status: str           # "healthy" | "degraded" | "unhealthy"
    last_heartbeat: str   # ISO 8601
    uptime_seconds: int = 0


@dataclass
class DiscoveryEntry:
    agent_id: str
    name: str
    owner: str
    capabilities: List[str]
    network: NetworkInfo
    health: HealthStatus
    registered_at: str        # ISO 8601
    metadata_uri: Optional[str] = None


class IAgentDiscovery(ABC):

    @abstractmethod
    async def register(self, entry: DiscoveryEntry) -> None:
        """Register an agent and its capabilities."""
        ...

    @abstractmethod
    async def unregister(self, agent_id: str) -> None:
        """Remove an agent from the discovery layer."""
        ...

    @abstractmethod
    async def query(self, capability: str) -> List[DiscoveryEntry]:
        """Find all agents that expose a given capability."""
        ...

    @abstractmethod
    async def list_all(self) -> List[DiscoveryEntry]:
        """List every registered agent."""
        ...

    @abstractmethod
    async def heartbeat(self, agent_id: str) -> None:
        """Emit a keep-alive so the registry knows the agent is alive."""
        ...

    async def find(self, capability: str) -> Optional[DiscoveryEntry]:
        """
        Return the best healthy agent for `capability`.
        Default implementation calls query() and picks the first healthy result.
        """
        entries = await self.query(capability)
        healthy = [e for e in entries if e.health.status == "healthy"]
        return healthy[0] if healthy else (entries[0] if entries else None)

    async def find_by_id(self, agent_id: str) -> Optional[DiscoveryEntry]:
        """
        Look up an agent by exact agent_id.
        Default implementation calls list_all() and filters.
        """
        for entry in await self.list_all():
            if entry.agent_id == agent_id:
                return entry
        return None

File: templates/typescript/interfaces/IAgentDiscovery.ts
/**
 * Discovery layer interface.
 * Implement this to plug into any backend:
 *   - LocalDiscovery   → in-memory (dev / testing)
 *   - HttpDiscovery    → REST-based registry
 *   - GossipDiscovery  → P2P gossip protocol
 *   - OnChainDiscovery → ERC-8004 on-chain registry
 */
export interface IAgentDiscovery {
  /**
   * Register an agent with its capability list.
   * `network` carries the agent's reachable address.
   */
  register(entry: DiscoveryEntry): Promise<void>;

  /**
   * Remove an agent from the discovery layer.
   */
  unregister(agentId: string): Promise<void>;

  /**
   * Find all agents that expose a given capability.
   * Returns a list of DiscoveryEntry objects.
   */
  query(capability: string): Promise<DiscoveryEntry[]>;

  /**
   * List every registered agent (no filter).
   */
  listAll(): Promise<DiscoveryEntry[]>;

  /**
   * Emit a heartbeat so the registry knows the agent is alive.
   */
  heartbeat(agentId: string): Promise<void>;

  /**
   * Find the best healthy agent for `capability`.
   * Convenience method — implementations should delegate to `query()`.
   */
  find?(capability: string): Promise<DiscoveryEntry | null>;

  /**
   * Look up an agent by exact agentId.
   * Convenience method — implementations should delegate to `listAll()`.
   */
  findById?(agentId: string): Promise<DiscoveryEntry | null>;
}

export interface DiscoveryEntry {
  agentId: string;
  name: string;
  owner: string;
  capabilities: string[];
  network: {
    protocol: 'http' | 'websocket' | 'grpc' | 'tcp' | 'libp2p';
    host: string;
    port: number;
    tls: boolean;
    peerId?: string;
    multiaddr?: string;
  };
  metadataUri?: string;
  health: {
    status: 'healthy' | 'degraded' | 'unhealthy';
    lastHeartbeat: string; // ISO 8601
    uptimeSeconds: number;
  };
  registeredAt: string; // ISO 8601
}
